From 0cb0af4360484477a32a991dec2586e6094fef35 Mon Sep 17 00:00:00 2001
From: Charles Ci-Jyun Wu <dminus@andestech.com>
Date: Tue, 7 Mar 2023 15:37:07 +0800
Subject: [PATCH 035/168] riscv: andes: Implement ioremap_wc for noncached
 memory remapping support

In the RISCV-Linux-5.4 branch, ioremap_nocache() is the function
that Andes uses to support noncache memory mapping.

But the Linux kernel ioremap_nocache() has been deprecated.
commit 4bdc0d676a64 ("remove ioremap_nocache and devm_ioremap_nocache")

And the generic RISCV ioremap() flow has also changed.
commit 38af57825313 ("riscv: use the generic ioremap code")

Signed-off-by: Charles Ci-Jyun Wu <dminus@andestech.com>
---
 arch/riscv/include/asm/io.h | 11 +++++++++++
 arch/riscv/mm/Makefile      |  3 +++
 arch/riscv/mm/ioremap.c     | 27 +++++++++++++++++++++++++++
 3 files changed, 41 insertions(+)
 create mode 100644 arch/riscv/mm/ioremap.c

diff --git a/arch/riscv/include/asm/io.h b/arch/riscv/include/asm/io.h
index 42497d487a17..f90ba568ed19 100644
--- a/arch/riscv/include/asm/io.h
+++ b/arch/riscv/include/asm/io.h
@@ -133,6 +133,17 @@ __io_writes_outs(outs, u64, q, __io_pbr(), __io_paw())
 #define outsq(addr, buffer, count) __outsq(PCI_IOBASE + (addr), buffer, count)
 #endif
 
+#ifdef CONFIG_MMU
+#ifdef CONFIG_ARCH_ANDES
+extern void __iomem *ioremap_wc(phys_addr_t phys_addr, size_t size);
+#define ioremap_wc ioremap_wc
+#define ioremap_wt ioremap_wc
+
+extern void iounmap(volatile void __iomem *addr);
+#define iounmap iounmap
+#endif /* CONFIG_ARCH_ANDES */
+#endif /* CONFIG_MMU */
+
 #include <asm-generic/io.h>
 
 #ifdef CONFIG_MMU
diff --git a/arch/riscv/mm/Makefile b/arch/riscv/mm/Makefile
index 3a4dfc8babcf..44ca0226adcf 100644
--- a/arch/riscv/mm/Makefile
+++ b/arch/riscv/mm/Makefile
@@ -18,6 +18,9 @@ obj-y += cacheflush.o
 obj-y += context.o
 obj-y += pgtable.o
 obj-y += pmem.o
+ifeq ($(CONFIG_ERRATA_ANDES),y)
+obj-y += ioremap.o
+endif
 
 ifeq ($(CONFIG_MMU),y)
 obj-$(CONFIG_SMP) += tlbflush.o
diff --git a/arch/riscv/mm/ioremap.c b/arch/riscv/mm/ioremap.c
new file mode 100644
index 000000000000..6347fc6c9d76
--- /dev/null
+++ b/arch/riscv/mm/ioremap.c
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2023 Andes Technology Corporation.
+ */
+#include <linux/module.h>
+#include <linux/export.h>
+#include <linux/io.h>
+#include <linux/pgtable.h>
+
+void __iomem *ioremap_wc(phys_addr_t phys_addr, size_t size)
+{
+	void __iomem *ret;
+	pgprot_t prot;
+
+	prot = pgprot_writecombine(PAGE_KERNEL);
+
+	ret = ioremap_prot(phys_addr, size, pgprot_val(prot));
+
+	return ret;
+}
+EXPORT_SYMBOL(ioremap_wc);
+
+void iounmap(volatile void __iomem *addr)
+{
+	generic_iounmap(addr);
+}
+EXPORT_SYMBOL(iounmap);
-- 
2.34.1

